### PR TITLE
Restructure semantic highlighting and responses

### PR DESCRIPTION
**Pull request by st0012**
### Motivation

- We want semantic highlighting to ve extensible in a way that addons can access to individual tokens without being able to erase the entire token stack.
- We're exploring ways to standardise responses that expose different APIs.

#### Dedicated `ResponseBuilder` classes

All extensible requests should have a specialised `ResponseBuilder` class, which should:

- Encapsulate the core data and only expose limited APIs to addons
- Be responsible for converting the internal data structure to interface objects.
     - So `SemanticHighlighting#response` should return `Interface::SemanticTokens` instead of the custom `SemanticToken` objects.

And instead of requiring individual response builders, I think it's great to mimic the autoloading configuration of requests.

#### Extensible SemanticHighlighting request

Now `Requests::SemanticHighlighting` also calls `create_semantic_highlighting_listener` on addons to create additional listeners.

### CODE REVIEW

File: lib/ruby_lsp/responses/response.rb
Comment (st0012) [inline lib/ruby_lsp/responses/response.rb:15]: Maybe the names should be these instead?

- `ResponseBuilders`
- `ResponseBuilder`
- `ResponseBuilder#response`

File: lib/ruby_lsp/requests/semantic_highlighting.rb
Comment (aryan-soni) [inline lib/ruby_lsp/requests/semantic_highlighting.rb:28]: The `extend T::Generic` is no longer necessary if we remove the declaration of the generic type

File: lib/ruby_lsp/response_builders/response.rb
Comment (aryan-soni) [inline lib/ruby_lsp/response_builders/response.rb:1]: The file name should probably be `response_builder.rb`

File: lib/ruby_lsp/response_builders/response.rb
Comment (aryan-soni) [inline lib/ruby_lsp/response_builders/response.rb:12]:
```suggestion
      ResponseType = type_member { { upper: Object } }
```

**Review by aryan-soni — COMMENTED** [review]
Took a look with @vinistock 

He'll follow up with a couple more comments, but looks pretty good, just a couple things we wanted to note

File: test/requests/semantic_highlighting_expectations_test.rb
Comment (vinistock) [inline test/requests/semantic_highlighting_expectations_test.rb:84]: Let's add a test for an addon that needs to override the token type rather than changing the modifiers and check what happens.

We do this for the Sorbet annotations and it'll be necessary for other addons. The base Ruby LSP pushes the `method` token type for method calls, but we want the annotations to have the `type` token type.

It's important we stick to the types and modifiers from the [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokenTypes) because that's what themes use to define the colours they will use. If we deviate from that, then existing themes will highlight code incorrectly.

The main question here is: if the Ruby LSP and an addon push different token types for the same exact location, does VS Code prefer the first or the second token?

If it prefers the second one, then maybe we don't need to do anything and just allow the last addon to have the final say. If it picks the first one, then we'll need to override the type.

Comment (st0012) [reply test/requests/semantic_highlighting_expectations_test.rb]: > Is there ever a case where an addon would like to modify a token that isn't the current one (last one we pushed)?

I can't think of any at the moment.

> And how would the addon author know they need to modify an existing token rather than adding a separate one? Is there an API we can provide that will guide them to do the right thing?

I was thinking about not adding `add_token` to the builder and use `instance_variable_get` in our own listener to insert tokens. Then the only exposed APIs are `replace_type` and `replace_modifier`. WDYT?

Comment (vinistock) [reply test/requests/semantic_highlighting_expectations_test.rb]: I'm not sure that would be enough. An addon could want to push a token that doesn't already exist.

What do you think about exposing an API and handling this complexity from our side? Something like this

```ruby
# This is pseudo code, but you get the idea
class SemanticHighlightingResponseBuilder
  def add_token(location, ...)
    if last_token.location == location
      # modify last token with type and modifiers
    else
      # push new token
    end
  end
end
```

Comment (st0012) [reply test/requests/semantic_highlighting_expectations_test.rb]: > An addon could want to push a token that doesn't already exist.

Do you have a use case for it? If not, maybe we can wait for feedback before introducing new APIs that ourselves would not need?

> What do you think about exposing an API and handling this complexity from our side?

It's worth exploring. But again, without a real use case I can not confidently say it'd be sufficient or necessary. For the ideas I have for `ruby-lsp-rails` and `ruby-lsp-rspec`, being able to replace type or modifier is already enough. Maybe when ERB support is shipped we can have new tokens? But if that's the case, let's design the API when that happens.

Also, I think we can have this discussion in a follow up issue/PR as the builder structure is needed to unblock other requests' refactors.

**Review by vinistock — APPROVED** [review]

**Pull request merged**
